Add support of extra mongodb collstats metrics

stefans-elastic opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
-->

## Proposed commit message

Add support of extra mongodb collstats metrics (size, count, avgObjSize, storageSize, totalIndexSize, totalSize, max, nindexes)

<!-- Mandatory
Explain here the changes you made on the PR.

Please explain:

- WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.
- WHY:  the rationale/motivation for the changes

This text will be pasted into the squash dialog when the change is committed and will be
a long term historical record of the change to help future contributors understand the
change, please help them by making it clear and comprehensive, they may be you.

If the commit title is adequate to describe both of these things, The text here may be omitted
or replaced with "See title". The title of the PR will be used as the commit message title when
the merge is made and the "See title" marker will be removed if present.

The text here and the PR title will be subject to the PR review process.
-->

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [ ] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).
- [ ] I have verified that any added dashboard complies with Kibana's [Dashboard good practices](https://docs.elastic.dev/ux-guidelines/data-viz/dashboard-good-practices) 

## Author's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- [ ] 

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Supersedes #123
-->
- Closes #12406

## Screenshots

<!-- Optional
Add here screenshots presenting:
- Kibana UI forms presenting configuration options exposed by the integration
- dashboards with collected metrics or logs
-->

<img width="1512" alt="Screenshot 2025-01-21 at 1 32 08 PM" src="https://github.com/user-attachments/assets/ef4adae2-72f9-4755-954c-4977b81cd728" />

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/20709)
* Commit: 61a706d5506e653e35127e5b3c984bfa7c094daf

cc @stefans-elastic

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/20709","number":20709,"commit":"61a706d5506e653e35127e5b3c984bfa7c094daf"}],"number":20709}
buildkite-pr-comment-->

gpop63 commented on packages/mongodb/data_stream/collstats/fields/fields.yml:
```diff
@@ -110,3 +110,38 @@
+          type: long
+          description: >
+            The average size of an object in the collection (in bytes).
+        - name: stats.storageSize
+          type: long
+          description: >
+            The total amount of storage allocated to this collection for document storage (in bytes).
```
You could add `metric_type: gauge` and `unit: byte`. Also field namings could be improved check naming conventions: https://www.elastic.co/guide/en/beats/devguide/current/event-conventions.html

stefans-elastic replied:
>> You could add metric_type: gauge and unit: byte.
is this only for `stats.storageSize` property?

>> Also field namings could be improved
don't names have to be the same as they are in `beats`? Also the names match naming of the fields returned by mongo's collstats command - https://www.mongodb.com/docs/manual/reference/command/collStats/#output

stefans-elastic commented:
> @stefans-elastic not sure in which release this feature ended up in metricbeat but you will have to bump the kibana version to that one.

I don't think `beats` part got released already (maybe I didn't need to be so quick with this PR. Do you think I should move this PR to `Draft` until `beats` gets a new release?

shmsr commented:
> > @stefans-elastic not sure in which release this feature ended up in metricbeat but you will have to bump the kibana version to that one.
> 
> I don't think `beats` part got released already (maybe I didn't need to be so quick with this PR. Do you think I should move this PR to `Draft` until `beats` gets a new release?

@stefans-elastic Your changes will be able in next minor release of beats i.e., 8.18. So, if you set kibana.version in top-level manifest.yml correctly, then you can also merge it now. This change will only be available when some user installs the 8.18 or above stack; and should choose to install this specific version of package.

ishleenk17 commented on packages/mongodb/changelog.yml:
```diff
@@ -1,3 +1,8 @@
+- version: "1.16.1"
```
1.17.0 since its an enhancement

stefans-elastic replied:
done

ishleenk17 commented on packages/mongodb/data_stream/collstats/fields/fields.yml:
```diff
@@ -110,3 +110,38 @@
+    - name: stats
+      type: group
+      fields:
+        - name: stats.size
```
If the group has stats, the field name can be size, count. We need not append stats.* to the field name for a stats group.

stefans-elastic replied:
done

ishleenk17 commented on packages/mongodb/manifest.yml:
```diff
@@ -1,6 +1,6 @@
 name: mongodb
 title: MongoDB
-version: "1.16.0"
+version: "1.16.1"
```
We should update the Kibana version as well to the version from which the change is available in beats.
Is the change backported to 8.17.x and 8.18 ? We should change Kibana version accordingly

stefans-elastic replied:
updated

ishleenk17 commented:
> > @stefans-elastic not sure in which release this feature ended up in metricbeat but you will have to bump the kibana version to that one.
> 
> I don't think `beats` part got released already (maybe I didn't need to be so quick with this PR. Do you think I should move this PR to `Draft` until `beats` gets a new release?

8.18 release is quite late. 
We can backport it to 8.17.x as well and then we will be able to merge this PR sometime next month.

stefans-elastic commented:
> > > @stefans-elastic not sure in which release this feature ended up in metricbeat but you will have to bump the kibana version to that one.
> > 
> > 
> > I don't think `beats` part got released already (maybe I didn't need to be so quick with this PR. Do you think I should move this PR to `Draft` until `beats` gets a new release?
> 
> 8.18 release is quite late. We can backport it to 8.17.x as well and then we will be able to merge this PR sometime next month.

trying to backport to 8.17 (https://github.com/elastic/beats/pull/42358). I think we should be able to include this in 8.17.2 release

stefans-elastic commented on packages/mongodb/manifest.yml:
```diff
@@ -14,7 +14,7 @@ icons:
 conditions:
   kibana:
-    version: "^8.13.0"
+    version: "^8.17.2 || ^9.0.0"
```
I've noticed that quite a few packages have `|| ^9.0.0` in the condition. Correct me if we don't need it in this package

ishleenk17 replied:
> I've noticed that quite a few packages have `|| ^9.0.0` in the condition. Correct me if we don't need it in this package

We are doing a bulk change of adding support of 9.0.0 . So you can remove that from this PR.

ishleenk17 replied:
> Even if the new fields are only available from 8.17, they probably don't do any harm on 8.13, and this way you can more easily keep maintaining support for versions from 8.13 to 8.17.

@jsoriano : In case we don't upgrade to 8.17.2, we will have new fields which actually don't exist for stack version lower than this but will still be shown in the README of the Integration. The user sitting on 8.13 might then also expect those fields which can't be supported. WDYT ?

stefans-elastic replied:
> We are doing a bulk change of adding support of 9.0.0 . So you can remove that from this PR.

removed 9.0.0 from the PR

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/20804)
* Commit: f8a682b5e05d2d61b3e46e28d9eab756c316522c

### Failed CI Steps
* [Check integrations mongodb](https://buildkite.com/elastic/integrations/builds/20804#01948e52-0dd1-438a-b238-c2706c7f42af)

### History
* :green_heart: [Build #20709](https://buildkite.com/elastic/integrations/builds/20709) succeeded 61a706d5506e653e35127e5b3c984bfa7c094daf

cc @stefans-elastic

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/20804","number":20804,"commit":"f8a682b5e05d2d61b3e46e28d9eab756c316522c"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/20709","number":20709,"commit":"61a706d5506e653e35127e5b3c984bfa7c094daf"}],"number":20804}
buildkite-pr-comment-->

stefans-elastic commented:
> Can we add a commit message in the PR description to maintain PR hygiene?

added

mrodm commented on packages/mongodb/manifest.yml:
```diff
@@ -1,6 +1,6 @@
 name: mongodb
 title: MongoDB
-version: "1.16.0"
+version: "1.17.0-next"
```
It's not allowed to publish "next" versions. 

I cannot find the documentation, now but it was here the steps previously
https://github.com/elastic/integrations/pull/9796/files#diff-7b9748c4f848852d8b40b331e43f9fd58560e923d065183d0eb3e5916d831c9bL146

Should it be set again to 1.16.0 here (keeping the changelog "-next" entry)? @ishleenk17 @jsoriano

mrodm replied:
The same happened in this PR https://github.com/elastic/integrations/pull/12407

That new version of the "iis" package was not published, since it is not allowed to publish "-next" versions.

stefans-elastic replied:
to  clarify: in manifest it should be `1.16.0` and in changelog it should be `1.17.0-next`, right?

ishleenk17 replied:
> The same happened in this PR #12407
> 
> That new version of the "iis" package was not published, since it is not allowed to publish "-next" versions.
> 
> Maybe you would need to update the manifest also for that other package ("iis" package).
> 
> WDYT ? @ishleenk17 @jsoriano

You wanted to point out to another PR here for IIS  @mrodm ?

ishleenk17 replied:
If we use next in this PR and merge it. I don't think we will be able to publish any other change to the mongodb package until this change is published.
There are other mongodb changes ongoing in parallel which would be merged soon. 
So, IMO lets wait to push this PR until release of 8.17.2.

stefans-elastic replied:
I've removed `-next`

jsoriano replied:
Why does this change need 8.17.2? It is only adding support for more fields, it should be safe to use in older versions.

ishleenk17 replied:
@stefans-elastic : Lets keep the version as 8.13.
We should figure out a way to depict in the fields on how to showcase that a particular set of fields are available only from 8.17.2 onwards. 
But the kibana version can remain unchanged.

stefans-elastic replied:
@ishleenk17 done

jsoriano commented on packages/mongodb/changelog.yml:
```diff
@@ -1,3 +1,8 @@
+- version: "1.17.0-next"
```
Appending the `-next` version is an approach that can be used to introduce a change without publishing a version. This can be used to delay the decision on publishing a package, for example if a set of related changes are introduced on different PRs.

This is not linked to the stack version, a package can be published for an stack version that doesn't exist yet.

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/20840)
* Commit: 09a1f18d6697de22d75497a9a3606546f4b0508a

### Failed CI Steps
* [Check integrations mongodb](https://buildkite.com/elastic/integrations/builds/20840#01949268-2c7d-47c9-9c47-350ddb2849f8)

### History
* :broken_heart: [Build #20804](https://buildkite.com/elastic/integrations/builds/20804) failed f8a682b5e05d2d61b3e46e28d9eab756c316522c
* :green_heart: [Build #20709](https://buildkite.com/elastic/integrations/builds/20709) succeeded 61a706d5506e653e35127e5b3c984bfa7c094daf

cc @stefans-elastic

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/20840","number":20840,"commit":"09a1f18d6697de22d75497a9a3606546f4b0508a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/20804","number":20804,"commit":"f8a682b5e05d2d61b3e46e28d9eab756c316522c"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/20709","number":20709,"commit":"61a706d5506e653e35127e5b3c984bfa7c094daf"}],"number":20840}
buildkite-pr-comment-->

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/20851)
* Commit: 25ff1e0114e4fa07d4033bdc558d6d00797a8bd8

### History
* :broken_heart: [Build #20840](https://buildkite.com/elastic/integrations/builds/20840) failed 09a1f18d6697de22d75497a9a3606546f4b0508a
* :broken_heart: [Build #20804](https://buildkite.com/elastic/integrations/builds/20804) failed f8a682b5e05d2d61b3e46e28d9eab756c316522c
* :green_heart: [Build #20709](https://buildkite.com/elastic/integrations/builds/20709) succeeded 61a706d5506e653e35127e5b3c984bfa7c094daf

cc @stefans-elastic

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/20851","number":20851,"commit":"25ff1e0114e4fa07d4033bdc558d6d00797a8bd8"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/20840","number":20840,"commit":"09a1f18d6697de22d75497a9a3606546f4b0508a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/20804","number":20804,"commit":"f8a682b5e05d2d61b3e46e28d9eab756c316522c"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/20709","number":20709,"commit":"61a706d5506e653e35127e5b3c984bfa7c094daf"}],"number":20851}
buildkite-pr-comment-->

stefans-elastic commented:
since the [PR](https://github.com/elastic/beats/pull/42358) in beats to backport this to 8.17 has been finally merged I think it is safe to assume the feature is going to be available in 8.17.2. So I think we can go ahead and approve/merge this PR (I'm pretty sure I've addressed all PR comments. If you notice something I haven't addressed - please let me know). WDYT?

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/21218)
* Commit: 34e661567989e0c8dc5915122a9275858da71c03

### History
* :green_heart: [Build #20851](https://buildkite.com/elastic/integrations/builds/20851) succeeded 25ff1e0114e4fa07d4033bdc558d6d00797a8bd8
* :broken_heart: [Build #20840](https://buildkite.com/elastic/integrations/builds/20840) failed 09a1f18d6697de22d75497a9a3606546f4b0508a
* :broken_heart: [Build #20804](https://buildkite.com/elastic/integrations/builds/20804) failed f8a682b5e05d2d61b3e46e28d9eab756c316522c
* :green_heart: [Build #20709](https://buildkite.com/elastic/integrations/builds/20709) succeeded 61a706d5506e653e35127e5b3c984bfa7c094daf

cc @stefans-elastic

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/21218","number":21218,"commit":"34e661567989e0c8dc5915122a9275858da71c03"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/20851","number":20851,"commit":"25ff1e0114e4fa07d4033bdc558d6d00797a8bd8"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/20840","number":20840,"commit":"09a1f18d6697de22d75497a9a3606546f4b0508a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/20804","number":20804,"commit":"f8a682b5e05d2d61b3e46e28d9eab756c316522c"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/20709","number":20709,"commit":"61a706d5506e653e35127e5b3c984bfa7c094daf"}],"number":21218}
buildkite-pr-comment-->

ishleenk17 reviewed (approved):
Looks good!
